File: my-code/0112-path-sum/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool ans=false;
    int target;
    void dfs(TreeNode* root,int sum)
    {
        if(root==NULL)
        {
            return;
        }
        if(root->left!=NULL) dfs(root->left,sum+root->left->val);
        else if(root->left==NULL) dfs(root->left,sum);
        if(root->right!=NULL) dfs(root->right,sum+root->right->val);
        else if(root->right==NULL) dfs(root->right,sum);
        if(root->left==NULL && root->right==NULL && target==sum)
        {
            ans=true;
        }
        return ;
    }
    bool hasPathSum(TreeNode* root, int targetSum) {
        if(root==NULL)
        {
            return false;
        }
        target=targetSum;
        dfs(root,root->val);
        return ans;
    }
};

File: my-code/0112-path-sum/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int target;
    Boolean ans=false;
    void dfs(TreeNode root,int sum)
    {
        if(root==null)
        {
            return;
        }
        if(root.left!=null) dfs(root.left,sum+root.left.val);
        else if(root.left==null) dfs(root.left,sum);
        if(root.right!=null) dfs(root.right,sum+root.right.val);
        else if(root.right==null) dfs(root.right,sum);
        if(target==sum && root.left==null && root.right==null)
        {
            ans=true;
        }
        return;
    }
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if(root==null)
        {
            return false;
        }
        target=targetSum;
        dfs(root,root.val);
        return ans;
    }
}
